compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -3872,49 +3871,39 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
 
-        val args = implicitArgs(wtp.paramInfos, 0, pt)
-
-        def propagatedFailure(args: List[Tree]): Type = args match {
-          case arg :: args1 =>
-            arg.tpe match {
-              case ambi: AmbiguousImplicits =>
-                propagatedFailure(args1) match {
-                  case NoType | (_: AmbiguousImplicits) => ambi
-                  case failed => failed
-                }
-              case failed: SearchFailureType => failed
-              case _ => propagatedFailure(args1)
-            }
-          case Nil => NoType
-        }
-
-        val propFail = propagatedFailure(args)
-
-        def issueErrors(): Tree = {
-          def paramSymWithMethodTree(paramName: TermName) =
-            if tree.symbol.exists then
-              tree.symbol.paramSymss.flatten
-                .map(sym => sym.name -> sym)
-                .toMap
-                .get(paramName)
-                .map((_, tree))
-            else
-              None
-
-          wtp.paramNames.lazyZip(wtp.paramInfos).lazyZip(args).foreach { (paramName, formal, arg) =>
-            arg.tpe match {
+        /** Reports errors for arguments of `appTree` that have a
+          * `SearchFailureType`, recursively traversing arguments that are
+          * themselves applications. `mt` must be the type of `appTree.fun`.
+          */
+        def reportErrors(appTree: Apply, mt: MethodType): Unit =
+          val Apply(fun, args) = appTree
+          for (paramName, formal, arg) <- mt.paramNames.lazyZip(mt.paramInfos).lazyZip(args) do
+            arg.tpe match
               case failure: SearchFailureType =>
-                report.error(
-                  missingArgMsg(arg, formal, implicitParamString(paramName, methodStr, tree), paramSymWithMethodTree(paramName)),
-                  tree.srcPos.endPos
-                )
-              case _ =>
-            }
-          }
-          untpd.Apply(tree, args).withType(propFail)
-        }
+                arg match
+                  case childAppTree: Apply =>
+                    childAppTree.fun.tpe.widen match
+                      case childMt: MethodType => reportErrors(childAppTree, childMt)
+                      case _ => ()
+                  case _ => ()
+
+                val methodStr = err.refStr(methPart(fun).tpe)
+                val paramStr = implicitParamString(paramName, methodStr, fun)
+                val paramSymWithMethodCallTree =
+                  fun.symbol.paramSymss.flatten
+                    .find(_.name == paramName)
+                    .map((_, appTree))
+                val message = missingArgMsg(arg, formal, paramStr, paramSymWithMethodCallTree)
+                // Note: if the same error type appears on several trees, we
+                // might report it several times, but this is not a problem
+                // because only the first one will be displayed. We traverse in
+                // post-order, so that the most detailed message gets displayed.
```

Comment:
Can I rely on this? I observed this behavior but haven't dug into it.

Reply:
Yes, that's what's implemented.